api/client/inventory.go

```diff
@@ -0,0 +1,481 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/trace"
+	"github.com/gravitational/trace/trail"
+	log "github.com/sirupsen/logrus"
+)
+
+// DownstreamInventoryControlStream is the client/agent side of a bidirectional stream established
+// between teleport instances and auth servers.
+type DownstreamInventoryControlStream interface {
+	// Send attempts to send an upstream message. An error returned from this
+	// method either indicates that the stream itself has failed, or that the
+	// supplied context was canceled.
+	Send(ctx context.Context, msg proto.UpstreamInventoryMessage) error
+	// Recv accesses the incoming/downstream message channel.
+	Recv() <-chan proto.DownstreamInventoryMessage
+	// Close closes the underlying stream.
+	Close(...error)
+	// Done signals that the stream has been closed.
+	Done() <-chan struct{}
+	// Error checks for any error associated with stream closure (returns `nil` if
+	// the stream is open, or if the the stream was closed explicitly via the Close()
+	// method).
+	Error() error
```

Comment:
So this interface offers no way to tell whether the stream is still open or whether it was closed successfully? If so, that seems like a confusing API.

Could the `Close` method return an error and we can use that to indicate whether the close operation succeeded?